What Is a Platform Modifications (#301)
cncf/tag-app-delivery: platforms-whitepaper/v1alpha1/paper.md

## What is a platform

At its more basic, a platform is a **product** developed by a 
platform engineering team that is custom built and composed of supporting 
reusable capabilities and services designed to reduce the cognitive load and burden 
experienced by **internal** stakeholders of an enterprise organization 
while still safely shipping software at speed [[4]]. 

Note that while internal stakeholders typically refer to developers, they can 
refer to DevOps, SREs, FinOps, Customer Success, and anyone who is responsible 

> for software delivery and cloud native infrastructure in an enterprise organization [[2]].

Reviewer: I think we can skip this side note here for the sake of being concise.

Reviewer 2: Maybe this could be an entry/clarification in the glossary? We already have `Platform users`, so if we are missing a few examples we could add them there?

Reviewer 3: I think the glossary would be a good place for that. Either way, I would consider listing user roles like testers, operators, and managers as examples of users. Otherwise we would probably need to explain what is FinOps, what is Customer Success, or what is DevOps (this one in particular could explode very quickly 😊).

The goal of any platform should be to drive 
organizational efficiency and productivity by empowering internal stakeholders to 
focus on their key responsibilities while giving them the tools and self-service 

> necessary to drive their day to day work. In order to discover those needs, a platform as product

Reviewer: This sentence is a helpful transition, thanks! I like how you describe TVP as a response to meeting the most basic needs of users.

approach is needed which means starting off with building only what is necessary and then evolving 
the platform over time with user feedback and usage metrics. For example, a "platform" may start off as a wiki page 
with links to provision each required capability. Some [[3]] have referred to this principle as seeking
the "thinnest viable platform" that meets an enterprise's requirements. 

As the user needs of internal stakeholders evolve in an organization and platform owners get more used feedback
and success metrics, platform engineering teams in an enterprise must take ownership of delivering the required capabilities. 

> This typically leads into the evolution of the thinnest viable platform into an **internal platform** in cloud-native computing 

Reviewer: I don't think TVP _leads_ to internal platforms. Rather internal platforms start out as thinly as possible. TVP is really only applicable to an internal platform - an external platform by definition has to cater to many organizations' needs.

I've suggested TVP is related to "internal" but I'm still thinking :thinking: how to express that and work in "internal" here!

Reviewer 2: Just want to see if we are on the same page with TVP as not just how something starts, but how it continues. I believe the industry already has a term for starting out small and that would be MVP (Minimum Viable Product). My understanding is that TVP evolves that to say, in the world of platforms, it is best to stay thin and continue to add value over as many widely available tools as possible.

So to expand further (since i won't be around to clarify 😅:
A thin platform may start minimal (only covering a few use cases) but grow past minimal by covering many use cases while staying thin by continuously looking for ways to outsource capabilities that were originally innovative and custom but have since become commodity in the industry.

separate supporting capabilities from application-specific logic more than
previous paradigms. In cloud-like environments resources and capabilities are
often managed independently and integrated with custom business components; such
resources may include databases and object stores, Kubernetes clusters,
observability collectors and dashboards and more. An internal platform provides
these to enterprise teams in ways that make them easy to integrate in their
applications and systems without having to have a deep innate knowledge about the underlying
components, reducing cognitive load on internal stakeholders such as developers who can then focus 
on the required work to deliver business outcomes. 

For example, some of the use cases that can be met with an internal platform include: 

1. Developers of products or services can automatically request a complete 
   development environment to support iterative research and development of new 
   features. This includes spaces in relevant services such as Kubernetes clusters, 
   task runners and artifact storage, membership in designated teams, and publication of connection 
   info such as URLs and secrets.

> 1. Finops and other internal stakeholders get access to a dashboard for managing costs across business units with the ability to define chargeback rules without having to understand the underlying compute infrastructure 

Reviewer 2: IMO this behaviour is FinOps, not the person being FinOps. I also fear that FinOps is a bit of a buzz word. We are already buzzwording a lot in our core domain (and trying to explain those words as we go!), so I would stay away from buzzwords that do not explicitly move forward platform engineering.

Also, my read is that stakeholders in this case are budget holders which naturally all roll into the CFO office (which is that FinOps?).

```suggestion
1. Budget holders get access to a dashboard for managing costs across business units with the ability to define chargeback rules without having to understand the underlying compute infrastructure 
```

1. Operators of third-party products and services can automatically provision 
   spaces and supporting services to deploy and use those third-party products and 
   services

> These examples illuminate the definition of a platform according to Martin Fowler[[2]], 

Reviewer 2: This was a quote from Bottcher on Fowler's blog. I have confirmed we cite correctly in the original, but I would be uncomfortable with only citing Martin here.

In addition, I fear that without the lead in of the quote, this is a bit of a throw away rather than the punchy statement that it was originally provided as. Was there a specific goal in mind by removing the industry quotes from the start of this section (both Atlassian and Bottcher/Fowler)?

Reviewer 3: I had the same concern about the quote. The author of that article is Evan Bottcher, published on Martin Fowler's blog.

for software delivery and cloud native infrastructure in an enterprise organization [[2]].

The goal of any platform should be to drive 
organizational efficiency and productivity by empowering internal stakeholders to 
focus on their key responsibilities while giving them the tools and self-service 
necessary to drive their day to day work. In order to discover those needs, a platform as product
approach is needed which means starting off with building only what is necessary and then evolving 
the platform over time with user feedback and usage metrics. For example, a "platform" may start off as a wiki page 
with links to provision each required capability. Some [[3]] have referred to this principle as seeking
the "thinnest viable platform" that meets an enterprise's requirements. 

As the user needs of internal stakeholders evolve in an organization and platform owners get more used feedback
and success metrics, platform engineering teams in an enterprise must take ownership of delivering the required capabilities. 
This typically leads into the evolution of the thinnest viable platform into an **internal platform** in cloud-native computing 
paradigms. 

> Internal platforms are particularly relevant for cloud-native architectures because these

Reviewer 3: Small suggestion:

```suggestion
Internal platforms are particularly relevant for cloud-native architectures because they
```

## What is a platform

> At its more basic, a platform is an **integrated product** developed by a 

Reviewer 2: I really like these first two paragraphs as a way to introduce and expand on what internal means 🙌

> At its more basic, a platform is an **integrated product** developed by a 

Author: removed the word integrated here

platform engineering team that is custom built and composed of supporting 
capabilities and services designed to reduce the cognitive load and burden 

> experienced by **internal** stakeholders of an enterprise organization 

Reviewer: While a platform may indeed often be a product, I don't think that's _intrinsic_ to its definition. That is, you _could_ have a platform that's not a product; in fact I'd argue that we emphasize the idea of "platform as product" because it's really _not_ a product - it's an internal utility but for it to succeed it must be treated _like_ a product.

I'd prefer we stick with the original 3 paragraphs for this section.

> experienced by **internal** stakeholders of an enterprise organization 

Reviewer 2: This is a tough one because I think that our group wants to advocate for the platform to be considered a platform while in reality that is still quite far to the right on the adoption curve.

By adding it here on the first line we are surely setting a stake in the ground which could either be very good way to build a shared assumption with the reader or could risk alienating. I am quite torn 🤔

while still safely shipping software at speed [[4]]. 

Note that while internal stakeholders typically refer to developers, they can 
refer to DevOps, SREs, FinOps, Customer Success, and anyone who is responsible 
for software delivery and cloud native infrastructure in an enterprise organization. 

> Platforms have become a key focus in cloud-native computing, which typically
> separates supporting capabilities from application-specific logic more than
> previous paradigms. In such environments resources like Kubernetes clusters, 
> databases and object stores, message queues and brokers, observability collectors
> and dashboards,user directories and authentication systems, task runners and reconcilers and
> more are managed independently and integrated into applications running in
> containers and machines. However, this has introduced its own set of challenges as 
> developers for example are having to spend more time learning about infrastructure 
> instead of focusing on software and app delivery.
> 
> Therefore, the goal of the platform is to truly drive
> organizational efficiency and productivity by empowering internal stakeholders to 
> focus on their key responsibilities while giving them the tools and self-service 
> necessary to drive their day to day work. For developers for example, this could mean
> having the ability to deploy self-service environments for application development without 
> the overhead of understanding infrastructure components. Or for FinOps, this could
> mean being able to define chargeback rules without understanding getting too in depth of 
> how the metrics are actually collectedfrom cloud-native infrastructure. 

Reviewer 2: I found this paragraph a bit confusing and difficult to read in the previous version as well. My concerns are:
1. The number of examples is difficult to parse
2. The use of cloud native here doesn't seem to lead to the challenges proposed.

Here is an attempt to capture what I mean:

```suggestion
It is said that today, every company is a technology company. In addition, the expectations on the availability and scalability of that software are growing. This has led to an increase in pressure on teams to manage more complex infrastructure like queues, high availability data stores, edge caches and more. For additional resiliency and scalability, this infrastructure is mainly built and managed on different physical machines from the software that depends on it. In order to configure and manage these resources developers are having to spend more time learning about infrastructure instead of focusing on software and app delivery.

A platform focuses on centralising the configuration and management of this infrastructure in order to enable organizational efficiency and productivity by empowering internal stakeholders to focus on their key responsibilities while giving application teams the tools and self-service. A team of platform experts not only reduces common work demanded of product teams but also optimizes platform capabilities used in those 
products. A platform team also maintains a set of conventional patterns, knowledge and tools used broadly across the enterprise ; enabling developers to quickly contribute to other teams and products built on the same 
foundations. The shared platform patterns also allow embedding governance and controls in templates, patterns and capabilities. Finally, because platform teams corral providers and provide consistent experiences bespoke to their unique needs over their offerings, they enable efficient use of public clouds and service providers for foundational but undifferentiated capabilities such as databases, identity access, infrastructure operations, and app lifecycle.
```

You will notice I pulled in the last paragraph of the why section as I think that we may want that here instead of in the above section?

Reviewer: The implicit question this paragraph is trying to answer is _why_ have "platforms" suddenly become important? The answer posited is that cloud-native computing's distributed systems paradigms lead to a need for platforms of capabilities.

We could make this question and answer more clear, perhaps along the lines of "platforms have become critical for cloud-native computing **because** supporting capabilities are no longer in line with business logic..."

The other items you all mention may fit best in the Attributes section.

Author: fixed and included some notes about TVP and how that can evolve into an internal platform

Therefore according to Martin Fowler[[2]], **"a digital platform is a foundation of self-service APIs, tools, services, 
knowledge and support which are arranged as a compelling internal product."**

It is key to note that a platform is bespoke to an organization, supporting 
a unique set of users and business needs. Therefore, platforms are not to be confused 
as just an Internal Developer Portal (IDP) or Service Catalog [[5]]. While those interfaces 
can certainly be helpful for many organizations, ultimately platform engineering teams

> need to take a platform as product approach and think deeply about the business objectives 

Reviewer: It would be best for us to focus on defining CNCF's opinion on platforms rather than countering other opinions - there will be no end to that! Can we take out the references to IDPs and service catalog here and instead list them in the Capabilities section where they fit? That would be enough to imply that we don't believe they're themselves a platform.

of their organizations. At its very basic level, a platform could simply be a place where 
users go for basic actions like user management, technical docs, or incident management if that 
is what is most needed.
 

> However, just like any product, a platform can evolve over time to meet more complex needs of an organization. 

Reviewer 2: ```suggestion
However, just like any product, a platform must evolve over time to meet evolving needs of an organization. 
```

For example, some example use cases an enterprise can meet with a platform include the following:

1. Developers of products or services can automatically request a complete 
   development environment to support iterative research and development of new 
   features. This includes spaces in relevant services such as Kubernetes clusters, 
   task runners and artifact storage, membership in designated teams, and publication of connection 
   info such as URLs and secrets.

> 1. Engineering Teams and SRE Teams have a central location for observability where infrastructure

Reviewer 2: I'd rather not get into more specific team names (e.g. SRE here)

> 1. Engineering Teams and SRE Teams have a central location for observability where infrastructure

Reviewer: Let's agree on a set of use cases together and then write them up to revise this list.

Author: addressed based on comments in last meeting - happy to talk through this more

	and services can be onboarded into a single observability platform. 
	This allows for a consistent approach to observability while giving developers 
	an easy interface to be able to do this for any app or cloud service.
1. Operators of third-party products and services can automatically provision 
   spaces and supporting services to deploy and use those third-party products and 
   services
   

> Therefore, It is recommended that platform engineering teams start with the thinnest viable platform (TVP) [[6]], 

Reviewer 2: Nice add of pointer to TVP

which is very similar to the concept of an MVP - solve
for the key needs of your organization with the most basic and continue to evolve your platform just like 

> any other product. 

Reviewer: Could we integrate this into a single paragraph on TVP and "internal"?

By offering consistent experiences for individual and/or scenario-oriented sets 
of capabilities, platforms make it easy for their users to deliver valuable products.

[1]: https://www.atlassian.com/devops/frameworks/team-topologies
[2]: https://martinfowler.com/articles/talk-about-platforms.html

> [3]: https://thenewstack.io/platform-as-a-product-true-devops/

Reviewer: I'm not yet sure how we'll do footnotes or if we want to highlight specific vendors and posts. In the meantime can we add these to the list at the end of the doc?
